<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="ur">
   <!-- import main macro -->
   <xacro:include filename="$(find ur_description)/urdf/inc/ur_macro.xacro"/>

   <!-- parameters -->
   <xacro:arg name="joint_limit_params" default=""/>
   <xacro:arg name="kinematics_params" default=""/>
   <xacro:arg name="physical_params" default=""/>
   <xacro:arg name="visual_params" default=""/>
   <!-- legal values:
         - hardware_interface/PositionJointInterface
         - hardware_interface/VelocityJointInterface
         - hardware_interface/EffortJointInterface
   -->
   <xacro:arg name="transmission_hw_interface" default="hardware_interface/PositionJointInterface"/>
   <xacro:arg name="safety_limits" default="false"/>
   <xacro:arg name="safety_pos_margin" default="0.15"/>
   <xacro:arg name="safety_k_position" default="20"/>

   <!-- arm -->
   <xacro:ur_robot
     prefix=""
     joint_limits_parameters_file="$(arg joint_limit_params)"
     kinematics_parameters_file="$(arg kinematics_params)"
     physical_parameters_file="$(arg physical_params)"
     visual_parameters_file="$(arg visual_params)"
     transmission_hw_interface="$(arg transmission_hw_interface)"
     safety_limits="$(arg safety_limits)"
     safety_pos_margin="$(arg safety_pos_margin)"
     safety_k_position="$(arg safety_k_position)"/>

   <link name="flange_interface">
     <visual>
       <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/FlangeInterface.stl" />
       </geometry>
       <material name="">
         <color rgba="1 1 1 1" />
       </material>
     </visual>
   </link>
   <joint name="flange-flange_interface" type="fixed">
     <parent link="flange" />
     <child link="flange_interface" />
     <origin xyz="0 0 0" rpy="0 0 0" />
   </joint>

   <link name="lower_ft_interface">
     <visual>
       <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/FTInterface.stl" />
       </geometry>
       <material name="">
         <color rgba="1 1 1 1" />
       </material>
     </visual>
   </link>
   <joint name="flange_interface-lower_ft_interface" type="fixed">
     <parent link="flange_interface" />
     <child link="lower_ft_interface" />
     <origin xyz="0.01 0 0" rpy="${pi/4.0} 0 0" />
   </joint>

   <link name="ati_ft_sensor">
     <visual>
       <!-- <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0" /> -->
       <origin xyz="0 0 0" rpy="0 0 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/ATI-9105-Axia80-M20.stl" />
       </geometry>
       <material name="">
         <color rgba="0.5 0.5 0.5 1" />
       </material>
     </visual>
   </link>
   <joint name="lower_ft_interface-ati_ft_sensor" type="fixed">
     <parent link="lower_ft_interface" />
     <child link="ati_ft_sensor" />
     <!-- <origin xyz="0.0354 0 0" rpy="${pi/2.0} 0 0" /> -->
     <origin xyz="0.0354 0 0" rpy="${pi} -${pi/2.0} 0" />
   </joint>

   <link name="upper_ft_interface">
     <visual>
       <origin xyz="0 0 0" rpy="0 -${pi/2.0} 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/FTInterface.stl" />
       </geometry>
       <material name="">
         <color rgba="1 1 1 1" />
       </material>
     </visual>
   </link>
   <joint name="ati_ft_sensor-upper_ft_interface" type="fixed">
     <parent link="ati_ft_sensor" />
     <child link="upper_ft_interface" />
     <!-- <origin xyz="0.01 0 0" rpy="${pi/2.0} 0 0" /> -->
     <origin xyz="0 0 0.01" rpy="0 -${pi/2} 0" />
   </joint>

   <link name="coupling_interface">
     <visual>
       <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/InterfaceToCoupling.stl" />
       </geometry>
       <material name="">
         <color rgba="1 1 1 1" />
       </material>
     </visual>
   </link>
   <joint name="upper_ft_interface-coupling_interface" type="fixed">
     <parent link="upper_ft_interface" />
     <child link="coupling_interface" />
     <origin xyz="0 0 0" rpy="0 0 0" />
   </joint>

   <link name="robotiq_coupling">
     <visual>
       <origin xyz="0 0 0" rpy="0 ${pi/2.0} 0" />
       <geometry>
         <mesh filename="package://robot_planning/resources/ISO9409-1-50-4-M6.stl" />
       </geometry>
       <material name="">
         <color rgba="0 0 0 1" />
       </material>
     </visual>
   </link>
   <joint name="coupling_interface-robotiq_coupling" type="fixed">
     <parent link="coupling_interface" />
     <child link="robotiq_coupling" />
     <origin xyz="0.01645 0 0" rpy="0 0 0" />
   </joint>

   <joint name="robotiq_coupling-robotiq_epick2cup_base_link" type="fixed">
     <parent link="robotiq_coupling" />
     <child link="robotiq_epick2cup_base_link" />
     <origin xyz="0.00845 0 0" rpy="0 0 -${pi/2}" />
   </joint>

   <!-- TODO Replace me with actual vacuum gripper URDF -->
   <xacro:include filename="$(find robot_planning)/urdf/robotiq_epick_2cups.urdf" />
</robot>
